Community/Tdarr_Plugin_DUrEX.js

```diff
@@ -0,0 +1,1000 @@
+    jobReport.steps.push(`target_pct_reduction_value: ${target_pct_reduction_value}`);
+    jobReport.steps.push(`target_bitrate: ${target_bitrate}`);
+
+    // Calculate the target bitrate based on reduction percentage and current bitrate
+    let bitratetarget;
+    if (bitrateprobe * target_pct_reduction_value > target_bitrate) {
+      bitratetarget = target_bitrate;
+      jobReport.steps.push(`Using target bitrate: ${bitratetarget} bps.`);
+    } else {
+      bitratetarget = bitrateprobe * target_pct_reduction_value;
+      jobReport.steps.push(`Using adjusted bitrate: ${bitratetarget} bps based on current bitrate and target_pct_reduction.`);
+    }
+
+    let bitratetarget_kbps = Math.round(bitratetarget / 1000); // Convert to kbps
+    jobReport.steps.push(`Target bitrate for output: ${bitratetarget_kbps} kbps`);
+
+    // Check for copy condition
+    if (stream.codec_name.toLowerCase() === 'hevc' && 
+        bitratetarget >= bitrateprobe * 0.95 && 
+        bitratetarget <= bitrateprobe * 1.05) {
+      jobReport.steps.push(`Copying video stream ${id} as codec and bitrate are unchanged.`);
+      configuration.AddOutputSetting("-c:v copy");
+      firstVideoStreamProcessed = true;
+      return;
+    }
+
+    let bufferSize_kbps = Math.round(bitratetarget_kbps * 2); // Set buffer size
+    jobReport.steps.push(`Buffer size: ${bufferSize_kbps} kbps`);
+
+    // build configuration based on hw encoder type
+
+	configuration.AddOutputSetting(`-c:v ${encoder.name} -qmin 0 -cq:v ${inputs.quality_target} -b:v ${bitratetarget_kbps}k -maxrate ${bitratetarget_kbps}k -bufsize ${bufferSize_kbps}k`);
+	jobReport.steps.push(`Transcoding to HEVC using ${encoder.name}`);
+
+
+    // If input is 4K and target_bitrate_4KUHD is set to 1, scale down to 1080p
+    if (inputs.target_bitrate_4KUHD === "1" && file.video_resolution === "4KUHD") {
+      configuration.AddOutputSetting(`-vf scale=1920:1080`); // Scale to 1080p
+      jobReport.steps.push("Adding resolution scaling to 1080p.");
+    }
+
+    firstVideoStreamProcessed = true; // Mark first video stream as processed
+  }
+
+  // Process all video streams
+  loopOverStreamsOfType(file, "video", videoProcess, jobReport);
+
+  if (!configuration.shouldProcess) {
+    jobReport.steps.push("No video processing necessary");
+  }
+
+  return configuration;
+}
+
+
+
+
+// #endregion
+
+const hasEncoder = async (encoder, ffmpegPath, jobReport) => {
+    const { exec } = require('child_process');
+    
+    return new Promise((resolve) => {
+        const command = `${ffmpegPath} -f lavfi -i color=c=black:s=256x256:d=1:r=30 -c:v ${encoder} -f null /dev/null`;
+        
+        exec(command, (error) => {
+            if (jobReport) {
+                jobReport.steps.push(`Checking encoder ${encoder}: ${error ? 'Not Available' : 'Available'}`);
+            }
+            resolve(!error);
+        });
+    });
+};
+
+const getBestHevcEncoder = async (otherArguments, jobReport) => {
```

Comment:
This is pretty opinionated, so it might be best to juste have a default list someone could re-order.

As in my opinion, it's CPU > QSV > NVENC > AMF, and I'm not exactly sure where Videotoolbox slots in.

Reply:
between nvenc amf
it does a pretty good job :)

100%
 CPU > QSV > NVENC > Videotoolbox > AMF,